core/src/main/java/com/linecorp/armeria/client/endpoint/WeightedRoundRobinStrategy.java

```diff
@@ -68,26 +72,112 @@ public Endpoint select(ClientRequestContext ctx) {
+            final long accumulatedWeight;
+
+            EndpointsGroupByWeight(long startIndex, int weight, long accumulatedWeight) {
+                this.startIndex = startIndex;
+                this.weight = weight;
+                this.accumulatedWeight = accumulatedWeight;
+            }
+        }
+
+        //
+        // In general, assume the weights are w0 < w1 < ... < wM where M = N - 1, N is number of endpoints.
+        //
+        // * The first part of result: (a0..aM)(a0..aM)...(a0..aM) [w0 times for N elements].
+        // * The second part of result: (a1..aM)...(a1..aM) [w1 - w0 times for N - 1 elements].
+        // * and so on
+        //
+        // In this way:
+        //
+        // * Total number of elements of first part is: X(0) = w0 * N.
+        // * Total number of elements of second part is: X(1) = (w1 - w0) * (N - 1)
+        // * and so on
+        //
+        // Therefore, to find endpoint for a sequence S = currentSequence % totalWeight, firstly we find
+        // the part which sequence belongs, and then modular by the number of elements in this part.
+        //
+        // Accumulation function F:
+        //
+        // * F(0) = X(0)
+        // * F(1) = X(0) + X(1)
+        // * F(2) = X(0) + X(1) + X(2)
+        // * F(i) = F(i-1) + X(i)
+        //
+        // We could easily find the part (which sequence S belongs) using binary search on F.
+        // Just find the index k where:
+        //
+        //                               F(k) <= S < F(k + 1).
+        //
+        // So, S belongs to part number (k + 1), index of the sequence in this part is P = S - F(k).
+        // Because part (k + 1) start at index (k + 1), and contains (N - k - 1) elements,
+        // then the real index is:
+        //
+        //                              (k + 1) + (P % (N - k - 1))
+        //
+        // For special case like w(i) == w(i+1). We just group them all together
+        // and mark the start index of the group.
+        //
         private static final class EndpointsAndWeights {
             private final List<Endpoint> endpoints;
             private final boolean weighted;
-            private final int maxWeight;
-            private final int totalWeight;
+            private final long totalWeight; // prevent overflow by using long
+            private final List<EndpointsGroupByWeight> accumulatedGroups;
 
             EndpointsAndWeights(Iterable<Endpoint> endpoints) {
+
+                // prepare immutable endpoints
+                this.endpoints = Streams.stream(endpoints)
+                        .filter(e -> e.weight() > 0) // only process endpoint with weight > 0
+                        .sorted(Comparator.comparing(Endpoint::weight)
+                                .thenComparing(Endpoint::host)
+                                .thenComparingInt(Endpoint::port))
+                        .collect(toImmutableList());
+                long numEndpoints = this.endpoints.size();
```

Comment:
nit: would you add `final` keyword to the variable which holds the value not to be changed like here?

Reply:
I would change and commit soon. Thank you very much.